pkg/bitrot/bitrot.go

```diff
@@ -0,0 +1,141 @@
+/*
+ * Minio Cloud Storage, (C) 2017 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package bitrot provides algorithms for detecting data corruption.
+package bitrot
+
+import (
+	"fmt"
+	"io"
+	"strconv"
+)
+
+// Algorithm identifies a bitrot detection algorithm that is implemented in another
+// package.
+type Algorithm uint
+
+const (
+	// Poly1305 specifies the poly1305 polynomial authenticator
+	Poly1305 Algorithm = iota
+
+	// GHASH specifies the GHASH (AES-GCM) polynomial authenticator
+	GHASH
+
+	// BLAKE2b512 specifies the BLAKE2b-512 hash function
+	BLAKE2b512
+
+	// SHA256 specifies the SHA256 hash function
+	SHA256
+
+	// add algorithms here
+
+	// UnknownAlgorithm indicates an unknown/invalid algorithm
+	UnknownAlgorithm
+)
+
+var keysizes = []int{
+	Poly1305:   32,
+	GHASH:      32,
+	BLAKE2b512: 0,
+	SHA256:     0,
+}
+
+var names = []string{
+	Poly1305:   "poly1305",
+	GHASH:      "ghash",
+	BLAKE2b512: "blake2b",
+	SHA256:     "sha256",
+}
+
+var hashes = make([]func([]byte) Hash, UnknownAlgorithm)
+
+// New returns a new Hash implementing the given algorithm.
+// It returns an error if the given key cannot be used by the algorithm.
+// New panics if the algorithm is unknown.
+func (a Algorithm) New(key []byte) (Hash, error) {
+	if keysize := a.KeySize(); keysize != len(key) {
+		return nil, fmt.Errorf("bitrot: bad keysize %d for algorithm #%s", len(key), a)
+	}
+	f := hashes[a]
+	if f != nil {
+		return f(key), nil
+	}
+	panic("bitrot: requested algorithm  #" + strconv.Itoa(int(a)) + " is not available")
+}
+
+// KeySize returns the size of valid keys for the given algorithm in bytes.
+// It panics if the algorithm is unknown.
+func (a Algorithm) KeySize() int {
+	if a >= 0 && a < UnknownAlgorithm {
```

Comment:
a is a `uint` so `a >= 0` is a NOP.
Also you can better test against len(keysizes) -- in case somebody would add it to the const array but forget to add it to the keysizes array.